### ISSUE TITLE
Collection of indexer related issues

### ISSUE THREAD

**Issue by daniel-dnil** [15knots/cmake4eclipse#103]
**This is a...**  

- [X] Bug Report  
- [ ] Enhancement Request

I've spend some time trying to summarize various indexer related issues with the cmake4eclipse plugin and made testcases to show each issue clearly. I don't know how these issue are related, so for now I keep them all in the same bugreport but feel free to split them out if needed.

Setup
=====
eclipse.buildId=4.10.0.I20181206-0815
CDT: 9.6.0.201812111211
cmake4eclipse: 1.13.0.201901031836
OS: Linux Mint 19
cmake version 3.10.2

Common Prepare
==============
git clone https://github.com/daniel-dnil/cm4e-tc.git && cd cm4e-tc

TC1: Bad provider information in sample project
==========================================
1) git checkout 7662d205

2) Import project C-1src from folder proj-cn into empty new WS, this is a copy from
   the cmake4eclipse-sample-projects repo.

3) Project will autobuild OK

4) Indexer runs after build, but fails to find any compiler built-in include paths (can't find stdio.h etc).
   C/C++ General->Preprocessor Include Paths, Macros etc.->Providers only contains
   "CMAKE_EXPORT_COMPILE_COMMANDS Parser". No "Includes" folder is shown in Eclipse project structure.

5) Close and re-open project

6) "Includes" folder is shown only after project close->reopen. Is this a CDT bug?

7) Add the "CMAKE_EXPORT_COMPILE_COMMANDS Compiler Built-ins" provider. Now the compiler built-in paths are found
   and the indexer runs OK. Commit this change to git, SHA 6c5a8c8. Now we have a state where the project can be
   built and open closed. Maybe CMAKE_EXPORT_COMPILE_COMMANDS Compiler Built-ins should be part of sample project?

TC2: language.settings.xml reverted upon project import
==================================================

Note: This is a duplicate of https://github.com/15knots/cmake4eclipse/issues/102 but I'm keeping it
here anyway since it has references to a testcase and steps to reproduce.

1) git checkout 6c5a8c8

2) Continue working in TC1 WS above, a file proj-cn/.settings/org.eclipse.cdt.managedbuilder.core.prefs is created with
   contents that is not clear how this is relevent to this plugin. Keep this file in the repository for this TC.
```
$ cat proj-cn/.settings/org.eclipse.cdt.managedbuilder.core.prefs
eclipse.preferences.version=1
environment/buildEnvironmentInclude/cdt.managedbuild.config.gnu.exe.debug.960404064/CPATH/delimiter=\:
environment/buildEnvironmentInclude/cdt.managedbuild.config.gnu.exe.debug.960404064/CPATH/operation=remove
environment/buildEnvironmentInclude/cdt.managedbuild.config.gnu.exe.debug.960404064/CPLUS_INCLUDE_PATH/delimiter=\:
environment/buildEnvironmentInclude/cdt.managedbuild.config.gnu.exe.debug.960404064/CPLUS_INCLUDE_PATH/operation=remove
environment/buildEnvironmentInclude/cdt.managedbuild.config.gnu.exe.debug.960404064/C_INCLUDE_PATH/delimiter=\:
environment/buildEnvironmentInclude/cdt.managedbuild.config.gnu.exe.debug.960404064/C_INCLUDE_PATH/operation=remove
environment/buildEnvironmentInclude/cdt.managedbuild.config.gnu.exe.debug.960404064/append=true
environment/buildEnvironmentInclude/cdt.managedbuild.config.gnu.exe.debug.960404064/appendContributed=true
```
3) Setup a new empty WS and import C-1src from folder proj-cn into this WS. Now
   project builds but the provider information is completely reverted back to CDT defaults ???
```
diff --git a/proj-cn/.settings/language.settings.xml b/proj-cn/.settings/language.settings.xml
index 7de4ced..89d2ff7 100644
--- a/proj-cn/.settings/language.settings.xml
+++ b/proj-cn/.settings/language.settings.xml
@@ -2,8 +2,13 @@
 <project>
        <configuration id="cdt.managedbuild.config.gnu.exe.debug.960404064" name="Debug">
                <extension point="org.eclipse.cdt.core.LanguageSettingsProvider">
-                       <provider-reference id="de.marw.cmake.cdt.language.settings.providers.CompileCommandsJsonParser" ref="shared-provider"/>
-                       <provider-reference id="de.marw.cmake.cdt.language.settings.providers.BuiltinsCompileCommandsJsonParser" ref="shared-provider"/>
+                       <provider copy-of="extension" id="org.eclipse.cdt.ui.UserLanguageSettingsProvider"/>
+                       <provider-reference id="org.eclipse.cdt.core.ReferencedProjectsLanguageSettingsProvider" ref="shared-provider"/>
+                       <provider-reference id="org.eclipse.cdt.managedbuilder.core.MBSLanguageSettingsProvider" ref="shared-provider"/>
+                       <provider class="org.eclipse.cdt.managedbuilder.language.settings.providers.GCCBuiltinSpecsDetector" console="false" env-hash="1036888353926601946" id="org.eclipse.cdt.managedbuilder.core.GCCBuiltinSpecsDetector" keep-relative-paths="false" name="CDT GCC Built-in Compiler Settings" parameter="${COMMAND} ${FLAGS} -E -P -v -dD &quot;${INPUTS}&quot;" prefer-non-shared="true">
+                               <language-scope id="org.eclipse.cdt.core.gcc"/>
+                               <language-scope id="org.eclipse.cdt.core.g++"/>
+                       </provider>
                </extension>
        </configuration>
 </project>
```
   I'm not completely sure, but this revert back to CDT defaults only seems to happen upon project import and it might be
   related to the proj-cn/.settings/org.eclipse.cdt.managedbuilder.core.prefs file being present in the project directory.
   It might also be some race condition since I think I've been able to have the language.settings.xml reset also 
   when the org.eclipse.cdt.managedbuilder.core.prefs is not present.

4) Close the Eclipse project, reset the language.settings.xml (git reset --hard HEAD) and re-open the project to
   get back to normal operation again. This is a very confusing problem when sharing the Eclipse projects in teams.

TC3: Project specific include folders scope issue 
=================================================

1) git checkout 219e723

2) Import two projects from folder proj-cn and proj-cn into empty new WS

3) Close both project and revert any changes done to language.settings.xml performed at import
   (due to issue in TC2 above).

4) Open both projects again. expand the "Includes" folder. Notice that the last include folder
   in the list is the same for both projects, which isn't correct. Each project should have an
   include folder matching the project name from the CMakeLists.txt. compile_command.json does
   contain the correct information, but what is displayed in the "Includes" folder isn't correct.

5) Close both project and restarts Eclipse. Open project "proj-cn" first and expand the "Includes"
   folder. This now shows the correct information. Next open project "proj-cx" and expand the
   "Includes" folder, it shows the same information as "proj-cn". This also works the other way
   around, seems like the information from the first project opened in the WS is used also for
   other projects in the WS.

6) Checking the entries from the CMAKE_EXPORT_COMPILE_COMMANDS Parser by selecting Properties
   on main.c and navigating to C/C++ General -> Preprocessor Include Paths, Macros etc. -> 
   Entries -> CMAKE_EXPORT_COMPILE_COMMANDS Parser shows the same problem that can be seen in
   the "Includes" folder of the project.

TC4: Project specific cmake variables scope issue
=================================================

1) git checkout d0f91b4

2) Import two projects from folder proj-cn and proj-cn into empty new WS

3) Close both project and revert any changes done to language.settings.xml performed at import
   (due to issue in TC2 above).

4) Close and restart Eclipse, open WS created in step 2) above

5) Open proj-cx _only_, check properties under C/C++ Build -> Cmake4eclipse -> Symbols. Note
   the dummy symbold PROJNAME=PROJ_CX.

6) Cancel out of the properties dialog for proj-cx and open proj-cn. Check properties under
   C/C++ Build -> Cmake4eclipse -> Symbols for proj-cn, note that same PROJNAME=PROJ_CX being
   present. This isn't true however, this symbol only exists in project proj-cx.

7) Close and restart Eclipse, open WS created in step 2) above

8) Perform step 5) and 6) again, but opening the project in the other order. Now 
   PROJNAME=PROJ_CX missing on both projects instead. Seems like there is an issue loading
   this information from the project file and scope for this dialog.

TC5: CMAKE_EXPORT_COMPILE_COMMANDS Compiler Built-ins scope issue
=================================================================

1) Make sure the arm-none-eabi-gcc cross compiler is installed on path
   This is easy in Ubuntu since there is a package (gcc-arm-none-eabi) for this

2) git checkout e395b25

3) Import two projects from folder proj-cn and proj-cn into empty new WS. proj-cx is now
   using a different toolchain (ARM cross compiler) compared to proj-cn

4) Close both project and revert any changes done to language.settings.xml performed at import
   (due to issue in TC2 above).

5) Close and restart Eclipse, open WS created in step 3) above

6) Open proj-cx first and check that the "Includes" folder now is showing paths
   to libc includes in the cross compiler locations.

7) Open proj-nc next and expand the "Includes" folder. Note that this project now not
   only contains the wrong include paths from the project specific setup but also the
   include paths from the cross compiler project "proj-cx".

8) In "proj-cn", select Properties on main.c and expand C/C++ General ->
   Preprocess Include Paths, Macros etc. -> Entries -> CMAKE_EXPORT_COMPILE_COMMANDS Compiler Built-ins
   Note that all "Built-ins" are from the cross compiler toolchain which isn't correct
   since proj-cn is still build using the native toolchain.

TC6: CMAKE_EXPORT_COMPILE_COMMANDS Compiler Built-ins flags issue
=================================================================

1) Keep the same WS as in TC5 above. Close all open projects in Eclipse

2) git checkout 93689e4

3) git clean -dfx to clean out all build files.

4) Open proj-cx which will now autobuild. Check compile_commands.json and notice the
   -std=gnu90 flag. This flag is set all the way back in the toolchain file for
   cmake and I expect this to also be present when the "Compiler Built-ins" provider
   runs

5) Check symbols again on proj-cx (like step 8) of TC5 above), notice __STDC_VERSION__
   not reflecting the C version set in the toolchain file. In my case set to 
   201112L. This isn't that much of an issue here with C in this simple project , but with C++ and STL it is 
   a major concern since the feature set of STL greatly depends on the C++ version and not 
   feeding the indexer with the correct information here makes the "Compiler Built-ins"
   provider more or less unusable with C++.

**Comment by 15knots** [comment]
Could you please split this into separate issues? Otherwise this issue will become unreadable, since comments on github issues work like a forum, but not as a mailing list.
Just use the **TC*** header from this issue as the new issue title, delete the text from the issue template and paste in the body of the **TC*** section.

**Comment by 15knots** [comment]
As for TC1: This should have  been reported in cmake4eclipse-sample-projects.
And, YRS, "CMAKE_EXPORT_COMPILE_COMMANDS Compiler Built-ins" provider was missing.

**Comment by daniel-dnil** [comment]
Ok, testcases and issues split into separate issues
